File: Day247/longestStringChain.go
package main 

import (
	"fmt"
	"sort"
)

// type Trie struct {
// 	next  [26]*Trie
// 	end bool
// 	branch  int
// }

// func maxInt(a int, b int) int {
// 	if (a > b) {
// 		return a
// 	}
// 	return b
// }

// func (trie *Trie) addWord(word string, pos int, branch int) {
// 	if pos  == len(word) {
// 		trie.end = true
// 		trie.branch = branch
// 		return
// 	} 
// 	temp := word[pos]
// 	if trie.next[temp-'a'] == nil {
// 		trie.next[temp-'a'] = &Trie{}
// 	}
// 	trie.next[temp-'a'].addWord(word, pos+1, branch)
// }

// func (trie *Trie) find (word string, pos int, flag bool) int {
// 	if pos == len(word) && flag {
// 		return trie.branch
// 	}
// 	max := 0
// 	for i:=0; i<26 && !flag; i++ {
// 		if trie.next[i] != nil {
// 			max = maxInt(max, trie.next[i].find(word, pos, true))
// 		}
// 	}
// 	if pos == len(word) {
// 		return max
// 	}
// 	temp := word[pos]
// 	if trie.next[temp-'a'] != nil {
// 		max = maxInt(max, trie.next[temp-'a'].find(word, pos+1, flag))
// 	}
// 	return max
// }

func longestStrChain(words []string) int {
	if len(words) == 1 {
		return 1
	}
	sort.Slice(words, func(i int, j int) bool {
		return len(words[i]) < len(words[j])
	})
	n := len(words)
	dp := make([]int, n) 
	for i:=0; i<n; i++ {
		dp[i] = 1
	}
	answer := 0
	for i:=1; i<n; i++ {
		start := sort.Search(i, func(j int) bool {
			return len(words[j]) >= len(words[i]) -1
		})
		for j:=start; j<i; j++ {
			if isPre(words[j], words[i]) {
				dp[i] = max(dp[i], dp[j]+1)
			}
		}
		answer = max(answer, dp[i])
	}
	return answer
}

func isPre (a string, b string) bool {
	if len(a)+1 !=len(b) {
		return false
	}
	count := 0
	for i,j :=0, 0; i<len(a) && j < len(b); {
		if a[i] != b[j] {
			if count == 0 {
				count ++
				j++
			} else {
				return false
			}
		} else {
			i++
			j++
		}
	}
	return true
}

func main() {
	arr := []string{"a", "b", "ba", "bca", "bda", "bdca"}
	fmt.Println(longestStrChain(arr))
}

File: Day247/longestStringChain.java
package Day247;

import java.util.Arrays;

class Trie {
    Trie[] next = new  Trie[26];
    boolean end = false;
    int branch = 0;
    void addWord(String word, int pos, int branch){
        if (pos == word.length()) {
            end = true;
            this.branch = branch;
            return;
        }
        char temp = word.charAt(pos);
        if (next[temp - 'a'] == null) {
            next[temp - 'a'] = new Trie();
        }
        next[temp-'a'].addWord(word, pos+1, branch);
    }
    public int find(String word, int pos, boolean flag) {
        if (pos == word.length() && flag) {
            return branch;
        } 
        int max = 0;
        for (int i=0; i<26 && !flag; i++) {
            if (next[i] == null) {
                continue;
            }
            max =Math.max(max, next[i].find(word, pos, true));
        }
        if (pos == word.length()) {
            return max;
        }
        char temp = word.charAt(pos);
        if (next[temp-'a'] != null) {
            max =Math.max(max, next[temp-'a'].find(word, pos+1, flag));
        }
        return max;
    }
    // @Override
    // public String toString() {
    //     return 
    // }
}

public class longestStringChain {

    public static int longestStrChain(String[] words) {
        int answer = 0;
        Arrays.sort(words, (a,b)-> b.length() - a.length());
        Trie trie = new Trie();
        for (int i=0; i<words.length; i++) {
            int current = 1 + trie.find(words[i], 0, false);
            trie.addWord(words[i], 0, current);
            answer = Math.max(answer, current);

        }
        return answer;
    }
    public static void main(String[] args) {
        String arr[] = {"a","b","ba","bca","bda","bdca"};
        System.out.println(longestStrChain(arr));
    }
}